### PR TITLE
plugin/transfer: notifies

### PR DESCRIPTION
**Pull request by chrisohaver**
### 1. Why is this pull request needed and what does it do?

Adds notifies to transfer plugin, per https://github.com/coredns/coredns/pull/3223#issuecomment-534099365

### 2. Which issues (if any) are related?
#3223 #3121

### 3. Which documentation changes (if any) need to be made?
included 

### 4. Does this introduce a backward incompatible change or deprecation?
no

### CODE REVIEW

File: plugin/transfer/README.md
Comment (miekg) [inline plugin/transfer/README.md]: s/Hosts/hosts/

File: plugin/transfer/README.md
Comment (miekg) [inline plugin/transfer/README.md]:
```suggestion
  exception off the `*` wildcard).
```

File: plugin/transfer/notifiy.go
Comment (miekg) [inline plugin/transfer/notifiy.go]: newline before this import

File: plugin/transfer/notifiy.go
Comment (miekg) [inline plugin/transfer/notifiy.go]: can you merge this with the notify function? Maybe name it, `notifyForZone` or something?

File: plugin/transfer/notifiy.go
Comment (miekg) [inline plugin/transfer/notifiy.go]: also, did you do a git move for this file? or just cut and paste?

File: plugin/transfer/setup.go
Comment (miekg) [inline plugin/transfer/setup.go]: how is this go routine terminated?

**Review by miekg — CHANGES_REQUESTED** [review]

File: plugin/transfer/notifiy.go
Comment (stp-ip) [reply plugin/transfer/notifiy.go]: Also the filename is broken: notifiy -> notify.go

Comment (chrisohaver) [reply plugin/transfer/notifiy.go]: > also, did you do a git move for this file? or just cut and paste?

IIRC, parts of it were copied (not cut or moved) from file/notify.   The file plugin will need to continue to implement it's own zone transfer implementation until it's deprecated and removed.

File: plugin/transfer/setup.go
Comment (chrisohaver) [reply plugin/transfer/setup.go]: IIUC, `t.Notify` will exit when the channel closes.  As is now, it would be up to the plugins that write to the channels to close their channel when stopping.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/transf..." ]
>> also, did you do a git move for this file? or just cut and paste?
>
>IIRC, parts of it we copied (not cut or moved) from file/notify.   The file plugin will need to continue to implement it's own zone transfer implementation until it's deprecated and removed.

true

File: plugin/transfer/setup.go
Comment (chrisohaver) [reply plugin/transfer/setup.go]: I have added a note in the Transferer interface Notify method docs to this effect.

File: plugin/transfer/notify.go
Comment (miekg) [inline plugin/transfer/notify.go]: please add a stop channel to be able to kill this go-routine. It's good form to have this ablity

File: plugin/transfer/notify.go
Comment (miekg) [inline plugin/transfer/notify.go]:
```suggestion
func (t Transfer) sendNotifies(zone string) error {
```

File: plugin/transfer/notify.go
Comment (miekg) [inline plugin/transfer/notify.go]: this functions never returns an error?

File: plugin/transfer/transfer.go
Comment (miekg) [inline plugin/transfer/transfer.go]: I wonder how this will work for the `auto` plugin? Can you provide an sample impl for that plugin? Zones come and go, so this channel may need to know the zone as well?

**Comment by mbrgm** [comment]
Any progress on this?

**Comment by chrisohaver** [comment]
> Any progress on this?

Just waiting for further review and approval.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/transf..." ]
>> Any progress on this?
>
>Just waiting for further review and approval.

*finds review*
*submits review*

sometimes github's interface is driving me crazy

File: plugin/transfer/notify.go
Comment (chrisohaver) [reply plugin/transfer/notify.go]: Yes, as written currently.  I can remove the error from the signature.

File: plugin/transfer/transfer.go
Comment (chrisohaver) [reply plugin/transfer/transfer.go]: The _transfer_ plugin associates `to` servers with the static zones defined in the transfer plugin block, or in the enclosing server block. For _auto_, the `to` servers would essentially be the same for all child zones.

Currently, `sendNotifies` does an exact zone match when determining which `to` servers to use for the notified zone.  That would prevent _auto_ from sending notifies for child zones, so I'll need to modify that logic to match sub zones, preferring longest match.

Comment (chrisohaver) [reply plugin/transfer/transfer.go]: 5d057df754699a02444d42a0a1b151a56d65bc72 modifies the logic to match sub zones, preferring longest match, to accommodate dynamically added sub zones in auto.

File: plugin/transfer/notify.go
Comment (chrisohaver) [reply plugin/transfer/notify.go]: > I can remove the error from the signature.

done in  9cd9080

File: plugin/transfer/notify.go
Comment (chrisohaver) [reply plugin/transfer/notify.go]: > please add a stop channel to be able to kill this go-routine. It's good form to have this ablity

added in 9cc95ebced0cbc50eceb95df9dddd5fb72da5de2

**Comment by chrisohaver** [comment]
> I wonder how this will work for the auto plugin? Can you provide an sample impl for that plugin?

I have a WIP branch containing a (completely untested first stab) implementation of transfer/notify for file and auto ... https://github.com/chrisohaver/coredns/tree/auto-file-notify.

**Comment by chrisohaver** [comment]
> > I wonder how this will work for the auto plugin? Can you provide an sample impl for that plugin?
> 
> I have a WIP branch containing a (completely untested first stab) implementation of transfer/notify for file and auto ... https://github.com/chrisohaver/coredns/tree/auto-file-notify.

OK, I have verified the [sample implementation](https://github.com/chrisohaver/coredns/tree/auto-file-notify) for _auto_ (and _file_) works with a local instance of bind acting as a secondary. Adding a new zone to the _auto_ directory triggers a notify to the secondary, and the secondary then makes a transfer request as expected.

Corefile used:
```
example:5300 {
    transfer {
        to 127.0.0.1:53
    }
    auto {
        directory auto/
    }
}
```

auto directory contained zones for foo.example, bar.example, etc

**Comment by chrisohaver** [comment]
resolved merge conflict in README, rebased with master.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/transf..." ]
>resolved merge conflict in README, rebased with master.

Thanks. I'll look at all these this Friday

**Comment by chrisohaver** [comment]
Note: the most recent commits (adding notify source option). Makes backward incompatible changes, primarily that you may only specify one remote server per `to` option.  I don't think this is _too_ big of a deal since nothing (in tree) actually implements the transfer plugin yet.

File: plugin/transfer/README.md
Comment (miekg) [inline plugin/transfer/README.md:19]: /HOST/ADDRESS

File: plugin/transfer/transfer.go
Comment (miekg) [inline plugin/transfer/transfer.go:59]: I think just exporting a function as done in https://github.com/coredns/coredns/pull/3972 is simpler, and side steps the whole ping. During setup you just grab the transfer plugin and then you call Notify() when ever you need it.

**Pull request closed without merging**
